Stackdriver: Support for SLO queries

sunker opened the pull request:
Problem description [here](https://github.com/grafana/grafana/issues/22627)

In this draft PR, I decided to go with option 2 that was described in the linked issue above. Furthermore, the SLO expression is built interactively, and not free text field as was discussed in the issue. 

Besides from a new query builder for SLOs, this PR includes: 
- Template variable support for SLO, i.e possibility to template SLO services, selectors and SLOs 
- Lots of refactoring. 
    - The QueryEditor is split up in two - MetricEditor and SLOEditor. 
    - The datasource.ts file is split up in two, datasource.ts and api.ts. api.ts handles rest requests and it also has a simple caching functionality. This refactoring made it possible to move the getProjects call from the backend to the frontend, which made the backend way cleaner than before. 
    - Lots of components has been rewritten from class components to functional components using hooks

sunker committed:
wip: add slo support

sunker committed:
Export DataSourcePlugin

sunker committed:
wip: break out metric query editor into its own component

sunker committed:
wip: refactor frontend - keep SLO and Metric query in differnt objects

sunker committed:
wip - load services and slos

sunker committed:
Fix broken test

sunker committed:
Add interactive slo expression builder

sunker committed:
Change order of dropdowns

sunker committed:
Refactoring backend model. slo unit testing in progress

sunker committed:
Unit test migration and SLOs

sunker committed:
Cleanup SLO editor

sunker committed:
Simplify alias by component

sunker committed:
Support alias by for slos

sunker committed:
Support slos in variable queries

sunker committed:
Fix broken last query error

sunker committed:
Update Help section to include SLO aliases

sunker committed:
streamline datasource resource cache

sunker committed:
Break out api specific stuff in datasource to its own file

sunker committed:
Move get projects call to frontend

sunker committed:
Refactor api caching

sunker committed:
Unit test api service

sunker committed:
Fix lint go issue

sunker committed:
Fix typescript strict errors

sunker committed:
Fix test datasource

sunker committed:
Use budget fraction selector instead of budget

sunker committed:
Reset SLO when service is changed

sunker committed:
Handle error in case resource call returned no data

papagian commented on pkg/tsdb/stackdriver/stackdriver.go:
```diff
@@ -152,43 +152,67 @@ func (e *StackdriverExecutor) buildQueries(tsdbQuery *tsdb.TsdbQuery) ([]*Stackd
+			params.Add("filter", buildFilterString(q.MetricQuery.MetricType, q.MetricQuery.Filters))
+			params.Add("view", q.MetricQuery.View)
+			setAggParams(&params, &q.MetricQuery, durationSeconds, query.IntervalMs)
+			sq.GroupBys = append(sq.GroupBys, q.MetricQuery.GroupBys...)
```
why group by should be passed as `aggregation.groupByFields` [parameter](https://github.com/grafana/grafana/blob/b8b85b08332f66646668a01e8368de8dae21329f/pkg/tsdb/stackdriver/stackdriver.go#L290) as well as in the `sq.GroupBys`?

sunker replied:
Good question! That's legacy stuff..I'll remove it! :)

sunker replied:
...or actually it has to be like that. params is just the URL params for the rest request, whereas sq.GroupBys is a slice that is used to build labels and return meta data.

papagian commented on pkg/tsdb/stackdriver/stackdriver.go:
```diff
@@ -152,43 +152,67 @@ func (e *StackdriverExecutor) buildQueries(tsdbQuery *tsdb.TsdbQuery) ([]*Stackd
+				params.Add("aggregation.perSeriesAligner", "ALIGN_MEAN")
+			} else {
+				params.Add("aggregation.perSeriesAligner", "ALIGN_NEXT_OLDER")
+			}
```
I assume that `aggregation.crossSeriesReducer` parameter is not meaningful for SLO queries?

sunker replied:
Correct!

papagian replied:
Not that important but it would be easier for comparing the two different query types to have a similar to `setAggParams` for SLO queries for setting `aggregation.perSeriesAligner` and `aggregation.alignmentPeriod` parameters.

sunker replied:
I agree! Broke out agg param building for SLOs into its own func

sunker committed:
Show real SLI display name

sunker committed:
Use unsafe prefix on will mount hook

sunker committed:
Store goal in query model since it will be used as soon as graph panel supports adding a threshold

sunker committed:
Add comment to describe why componentWillMount is used

papagian commented on pkg/tsdb/stackdriver/stackdriver_test.go:
```diff
@@ -208,6 +208,95 @@ func TestStackdriver(t *testing.T) {
+								"metricType": "a/metric/type",
+								"view":       "FULL",
+								"aliasBy":    "testalias",
+								"type":       "timeSeriesQuery",
```
It would be nice to include a `groupBys` field and check the `queries[0].Params["aggregation.groupByFields"]` and `queries[0].GroupBys`.

sunker replied:
👍

sunker committed:
Interpolate sloid

papagian commented on pkg/tsdb/stackdriver/types.go:
```diff
@@ -14,6 +14,42 @@ type (
+		AliasBy            string
+		MetricKind         string
+		ValueType          string
+		View               string
```
I don't think that `Unit`, `MetricKind` and `ValueType` are used.

sunker replied:
👍

daniellee reviewed (changes requested):
Mostly works great but found 3 bugs and 2 of them have to be fixed (view field not getting the default value set anymore and the template variable bug that hangs the browser) and the third is more cosmetic.

- Adding or editing a template variable hangs Chrome or Firefox on Ubuntu. If I edit an existing Stackdriver template variable, it hangs after a 2-3 seconds. If I add a new variable, as soon as I have chosen a Stackdriver datasource then it hangs before I can click on anything else. To recreate: create new dashboard, create new template variable then choose a Stackdriver datasource:
![image](https://user-images.githubusercontent.com/434655/77370595-82af4000-6d61-11ea-921d-b4c17ffbf83a.png)

- When a template variable is used in the alignment field then the variable name is not visible in the select instead you see "Select Alignment":
![image](https://user-images.githubusercontent.com/434655/77370109-6fe83b80-6d60-11ea-9c57-ab2ef746a5f3.png)

sunker committed:
Break out SLO aggregation into its own func

sunker committed:
Also test group bys for metricquery test

sunker committed:
Remove not used type fields

sunker committed:
Remove annoying stackdriver prefix from error message

sunker committed:
Default view param to FULL

sunker commented on pkg/tsdb/stackdriver/stackdriver.go:
```diff
@@ -152,43 +152,67 @@ func (e *StackdriverExecutor) buildQueries(tsdbQuery *tsdb.TsdbQuery) ([]*Stackd
+
+		if q.QueryType == metricQueryType {
+			params.Add("filter", buildFilterString(q.MetricQuery.MetricType, q.MetricQuery.Filters))
+			params.Add("view", q.MetricQuery.View)
```
Yes! Nice catch there - thought I could remove this now since it's being set in the frontend, but had forgotten about altering. Thanks!

sunker commented:
> Mostly works great but found 3 bugs and 2 of them have to be fixed (view field not getting the default value set anymore and the template variable bug that hangs the browser) and the third is more cosmetic.
> 
> * Adding or editing a template variable hangs Chrome or Firefox on Ubuntu. If I edit an existing Stackdriver template variable, it hangs after a 2-3 seconds. If I add a new variable, as soon as I have chosen a Stackdriver datasource then it hangs before I can click on anything else. To recreate: create new dashboard, create new template variable then choose a Stackdriver datasource:
>   ![image](https://user-images.githubusercontent.com/434655/77370595-82af4000-6d61-11ea-921d-b4c17ffbf83a.png)
> * When a template variable is used in the alignment field then the variable name is not visible in the select instead you see "Select Alignment":
>   ![image](https://user-images.githubusercontent.com/434655/77370109-6fe83b80-6d60-11ea-9c57-ab2ef746a5f3.png)

Thanks for feedback! Strangely, I can't reproduce the alignment period template variable bug and the browser freeze bug. Maybe you can help med test that in today's handover meeting?

papagian commented on public/app/plugins/datasource/stackdriver/types.ts:
```diff
@@ -21,39 +21,70 @@ export enum MetricFindQueryTypes {
+  selectorName: string;
+  serviceId: string;
+  sloId: string;
+  goal?: number;
```
What is `goal` for? I don't see a respective field in [SLOQuery](https://github.com/grafana/grafana/blob/b7d7d5b5fa993585fcb3fb85158cbfcd74cda8ff/pkg/tsdb/stackdriver/types.go#L37) in the backend.

sunker replied:
Correct! 
When defining an SLO you have to specify a goal, i.e a level that you don't want to go under. See the goal being displayed as a dotted line in Stackdriver. I wanted to save the goal in the query model so that we can implement this easily when the graph panel is converted to react and it has support for displaying dotted lines like this. The new graph panel will probably be released in a few months of time.
<img width="303" alt="Skärmavbild 2020-03-24 kl  10 01 40" src="https://user-images.githubusercontent.com/2388950/77407488-7e614200-6db6-11ea-8e5f-429eef0c2648.png">

daniellee replied:
Maybe a comment about that will make this more obvious for anyone else looking at the code?

sunker committed:
Add part about SLO query builder in docs

papagian commented:
This [comment](https://github.com/grafana/grafana/blob/b7d7d5b5fa993585fcb3fb85158cbfcd74cda8ff/pkg/tsdb/stackdriver/types.go#L9) is a bit misleading since `StackdriverQuery` is no more sent by the frontend but it's created in the backend.

sunker committed:
Use new images

sunker commented:
> This [comment](https://github.com/grafana/grafana/blob/b7d7d5b5fa993585fcb3fb85158cbfcd74cda8ff/pkg/tsdb/stackdriver/types.go#L9) is a bit misleading since `StackdriverQuery` is no more sent by the frontend but it's created in the backend.

I agree! Actually, none of these types are exported so they should be camel cased. Won't need that comment then. I'll fix!

papagian commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -86,58 +86,62 @@ If Grafana is running on a Google Compute Engine (GCE) virtual machine, it is po
+
+The metric query editor allows you to select metrics, group/aggregate by labels and by time, and use filters to specify which time series you want in the results.
 
 Begin by choosing a `Project`. Then select a `Service` and then a metric from the `Metric` dropdown. Use the plus and minus icons in the filter and group by sections to add/remove filters or group by clauses.
```
Before one selecting `Project` needs to select `QueryType`.

sunker replied:
👍

papagian commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -174,6 +178,19 @@ Example Alias By: `{{resource.type}} - {{metric.type}}`
+
+To get an understanding of the basic concepts in service monitoring, please refer to Google Stackdriver's [official docs](https://cloud.google.com/monitoring/service-monitoring).
+
+The SLO query builder in the Stackdriver data source allows you to display SLO data in time series format. Begin by choosing a `Project`. Then select a `Service` and a `SLO`. Use the `Selector` field to specify which time series selector to use.
```
Same as above, one needs to select `Query Type` first.

sunker replied:
👍

papagian commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -174,6 +178,19 @@ Example Alias By: `{{resource.type}} - {{metric.type}}`
+
+#### Time series selectors
+
+Detailed descriptions about all the avaiable time series selectors can be found [here](https://cloud.google.com/monitoring/service-monitoring/timeseries-selectors#ts-selector-list).
```
`avaiable` -> `available`

sunker replied:
👍

sunker committed:
Fixes after feedback

papagian commented on pkg/tsdb/stackdriver/stackdriver_test.go:
```diff
@@ -208,6 +208,98 @@ func TestStackdriver(t *testing.T) {
+				So(queries[0].Target, ShouldEqual, "aggregation.alignmentPeriod=%2B60s&aggregation.crossSeriesReducer=REDUCE_NONE&aggregation.groupByFields=metric.label.group1&aggregation.groupByFields=metric.label.group2&aggregation.perSeriesAligner=ALIGN_MEAN&filter=metric.type%3D%22a%2Fmetric%2Ftype%22&interval.endTime=2018-03-15T13%3A34%3A00Z&interval.startTime=2018-03-15T13%3A00%3A00Z&view=FULL")
+				So(len(queries[0].Params), ShouldEqual, 8)
+				So(queries[0].Params["aggregation.groupByFields"][0], ShouldEqual, "metric.label.group1")
+				So(queries[0].Params["aggregation.groupByFields"][1], ShouldEqual, "metric.label.group2")
```
In addition to this you can check also `queries[0].GroupBys`.

sunker replied:
👍

sunker committed:
Add one more group by test

papagian commented on public/app/plugins/datasource/stackdriver/components/AliasBy.tsx:
```diff
@@ -1,53 +1,25 @@
-    this.propagateOnChange = debounce(this.props.onChange, 500);
-    this.state = { value: '' };
-  }
+  const propagateOnChange = debounce(onChange, 1000);
```
Is `1000` intentional? It used to `500`.

sunker committed:
Make stackdriver types internal

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -86,58 +86,62 @@ If Grafana is running on a Google Compute Engine (GCE) virtual machine, it is po
+
+The metric query editor allows you to select metrics, group/aggregate by labels and by time, and use filters to specify which time series you want in the results.
+
+To create a metric query, begin by choosing the option `Metrics` in the `Query Type` dropdown. Then select a `Project`, a `Service` and then a metric from the `Metric` dropdown. Use the plus and minus icons in the filter and group by sections to add/remove filters or group by clauses.
```
Please rewrite this as numbered steps. Also, per the [Documentation style guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md) and [Microsoft style guide](https://docs.microsoft.com/en-us/style-guide/welcome/), UI items are generally formatted as bold, not code.

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -86,58 +86,62 @@ If Grafana is running on a Google Compute Engine (GCE) virtual machine, it is po
+##### Metadata labels
 
-Resource metadata labels contains information to uniquely identify a resource in Google cloud. Metadata labels are only returned in the time series response if they're part of the **Group By** segment in the time series request. There's no API for retrieving metadata labels, so it's not possible to populate the group by dropdown with the metadata labels that are available for the selected service ...
+Resource metadata labels contains information to uniquely identify a resource in Google cloud. Metadata labels are only returned in the time series response if they're part of the **Group By** segment in the time series request. There's no API for retrieving metadata labels, so it's not possible to populate the group by dropdown with the metadata labels that are available for the selected service ...
```
```suggestion
Resource metadata labels contain information to uniquely identify a resource in Google Cloud. Metadata labels are only returned in the time series response if they're part of the **Group By** segment in the time series request. There's no API for retrieving metadata labels, so it's not possible to populate the group by dropdown with the metadata labels that are available for the selected service and metric. However, the **Group By** field dropdown comes with a pre-defined list of common system labels.
```

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -86,58 +86,62 @@ If Grafana is running on a Google Compute Engine (GCE) virtual machine, it is po
 User labels cannot be pre-defined, but it's possible to enter them manually in the **Group By** field. If a metadata label, user label or system label is included in the **Group By** segment, then you can create filters based on it and expand its value on the **Alias** field.
 
-### Alias Patterns
+#### Alias Patterns
```
```suggestion
#### Alias patterns
```

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -174,6 +178,19 @@ Example Alias By: `{{resource.type}} - {{metric.type}}`
 
 Example Result: `gce_instance - compute.googleapis.com/instance/cpu/usage_time`
 
+### SLO Queries
```
```suggestion
### SLO queries
```

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -174,6 +178,19 @@ Example Alias By: `{{resource.type}} - {{metric.type}}`
+
+To get an understanding of the basic concepts in service monitoring, please refer to Google Stackdriver's [official docs](https://cloud.google.com/monitoring/service-monitoring).
+
+The SLO query builder in the Stackdriver data source allows you to display SLO data in time series format. To create a SLO query, begin by choosing the option `Serice Level Objectives (SLO)` in the `Query Type` dropdown. Then select a `Project`, a `Service` and a `SLO`. Use the `Selector` field to specify which time series selector to use.
```
Please rewrite this as numbered steps.

bergquist reviewed (approved):
Tested it and checked the backend code and it looks ready. 

I wonder if the **metricType** dropdown should be placed somewhere else. I would suspect that most (almost everyone) who uses the Stackdriver datasource want to issue a metric query. So asking them to pick type as first option might be weird. 

@daniellee Did you forward the template issue to the frontend platform team? I guess that is not a blocker for this PR anymore?

sunker committed:
Update docs/sources/features/datasources/stackdriver.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

sunker committed:
Update docs/sources/features/datasources/stackdriver.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

sunker committed:
Update docs/sources/features/datasources/stackdriver.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

sunker commented:
> I wonder if the metricType dropdown should be placed somewhere else. I would suspect that most (almost everyone) who uses the Stackdriver datasource want to issue a metric query. So asking them to pick type as first option might be weird.

Yes, I agree that's what most users will use. But `Metric` is the default value when adding a new query row, so they won't have to select it manuallt.

sunker committed:
Updates after PR feedback

daniellee reviewed (approved):
The two issues I found were due to me having a feature flag on for the new React template variable implementation so nothing to do with this PR.

LGTM

bergquist commented:
@oddlittlebird could you please review this again? Thanks! :)

papagian reviewed (approved):
Great work!

oddlittlebird commented on docs/sources/features/datasources/stackdriver.md:
```diff
@@ -86,58 +86,67 @@ If Grafana is running on a Google Compute Engine (GCE) virtual machine, it is po
+
+The metric query editor allows you to select metrics, group/aggregate by labels and by time, and use filters to specify which time series you want in the results.
+
+To create a metric query, follow these steps:
```
Please add periods at the end of each sentence in this procedure.

oddlittlebird reviewed (approved):
Needs a few periods, good to go aside from that.

daniellee committed:
add test for when no alias by defined

daniellee committed:
fix infinite loop when newVariables feature flag is on

onChange being called in componentDidUpdate produces an
infinite loop when using the new React template variable
implementation.

Also fixes a spelling mistake

daniellee committed:
implements feedback for documentation changes

daniellee committed:
more doc changes
